### PR TITLE
add dist training with restnet model

### LINKED ISSUE #1268: Add resnet model with distributed mode

### PR DESCRIPTION
**Pull request by Yancey0623**
fixed #1268

### CODE REVIEW

File: fluid/image_classification/dist_train/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Distributed Image Classification Models Training
+
+This folder contains implementations of **Image Classification Models**, they are designed to support
+large-scaled distributed training with two distributed architecture: parameter server with RPC and ring-base with Nvidia NCCL2 library.
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/README.md]: We may need to have a general name for these two modes, how about parameter server mode and NCCL2 collective mode?

File: fluid/image_classification/dist_train/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+## Getting Started
+
+Before getting started, please make sure you have finished the imagenet [Data Preparation](../README.md#data-preparation).
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/README.md]: finished -> go through

File: fluid/image_classification/dist_train/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+1. The entrypoint file is `dist_train.py`, some important flags are as follows:
+
+    - `model`, the model to run with, such as `ResNet50`, `ResNet101` and etc..
+    - `batch_size`, the batch_size per device.
+    - `update_method`, specify the update method, local, pserver or nccl2.
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/README.md]: can choose from "local", "pserver" or "nccl2".

File: fluid/image_classification/dist_train/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+    - `PADDLE_TRAINING_ROLE`, the current training role, should be in [PSERVER, TRAINER].
+    - `PADDLE_TRAINERS`, the trainer count of a job.
+    - `PADDLE_CURRENT_IP`, the current instance IP.
+    - `PADDLE_PSERVER_IPS`, the parameter server IP list, separated by ","  only be used with update_method is pserver.
+    - `PADDLE_TRAINER_ID`, the unique trainer ID of a job, the ranging is [0, PADDLE_TRAINERS).
+    - `PADDLE_PSERVER_PORT`, the port of the parameter pserver listened on.
+    - `PADDLE_TRAINER_IPS`, the trainer IP list, separated by ",", only be used with upadte_method is nccl2.
+
+### Pserver Server with RPC
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/README.md]: Same as above, need a general name

File: fluid/image_classification/dist_train/README.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    PADDLE_TRAINER_ID=0 \
+    python dist_train.py \
+        --model=ResNet50 \
+        --batch_size=32 \
+        --update_method=pserver \
+        --device=GPU \
+        --data_dir=../data/ILSVRC2012
+    ```
+
+### Training Curve
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/README.md]: Training Curve -> Visualize the Training Process

File: fluid/image_classification/dist_train/README.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+It's easy to draw the training curve accroding to the training logs, for example,
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/README.md]: learning curve

File: fluid/image_classification/dist_train/args.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+__all__ = ['parse_args', ]
+
+BENCHMARK_MODELS = [
+    "ResNet50", "ResNet101", "ResNet152"
+]
+
+
+def parse_args():
+    parser = argparse.ArgumentParser('Fluid model benchmarks.')
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/args.py]: "Fluid model benchmarks."?

File: fluid/image_classification/dist_train/args.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    parser.add_argument(
+        '--use_reader_op',
+        action='store_true',
+        help='Whether to use reader op, and must specify the data path if set this to true.'
+    )
+    parser.add_argument(
+        '--no_random',
+        action='store_true',
+        help='If set, keep the random seed and do not shuffle the data.')
+    parser.add_argument(
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/args.py:105]: Can remove args that not used for this case?

File: fluid/image_classification/dist_train/README.md (hunk 1 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/README.md]: Thanks, done.

File: fluid/image_classification/dist_train/README.md (hunk 2 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/README.md]: done.

File: fluid/image_classification/dist_train/README.md (hunk 3 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/README.md]: done.

File: fluid/image_classification/dist_train/README.md (hunk 4 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/README.md]: done.

File: fluid/image_classification/dist_train/README.md (hunk 5 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/README.md]: done.

File: fluid/image_classification/dist_train/README.md (hunk 6 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/README.md]: done.

File: fluid/image_classification/dist_train/args.py (hunk 7 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/args.py]: done.

File: fluid/image_classification/dist_train/args.py (hunk 8 above)
Comment (Yancey0623) [reply fluid/image_classification/dist_train/args.py:105]: done.

File: fluid/image_classification/dist_train/dist_train.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+import time
+import os
+import traceback
+
+import numpy as np
+
+import paddle
+import paddle.fluid as fluid
+import paddle.fluid.core as core
+import paddle.fluid.transpiler.distribute_transpiler as distribute_transpiler
```
Comment (typhoonzero) [inline fluid/image_classification/dist_train/dist_train.py]: can use `fluid.DistributeTranspiler` now.

**Review by typhoonzero — APPROVED** [review]
LGTM for now, still need to clean some code copied from fluid_benchmark.py, please @qingqing01 take a look at this

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**